core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java

```diff
@@ -86,8 +92,12 @@ public static ShapeBuilder parse(XContentParser parser)
     }
 
     /** parse geometry from the stream tokenizer */
-    private static ShapeBuilder parseGeometry(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
+    private static ShapeBuilder parseGeometry(StreamTokenizer stream, GeoShapeType shapeType)
+            throws IOException, ElasticsearchParseException {
         final GeoShapeType type = GeoShapeType.forName(nextWord(stream));
+        if (shapeType != null && type.wktName().equals(shapeType.wktName()) == false) {
```

Comment:
I think it would be good to have a small test for this kind of type check.

Reply:
:+1: